File: python/a29.py
# A29 Power
# submission: https://atcoder.jp/contests/tessoku-book/submissions/38574873

MODULO = 10 ** 9 + 7

def power(x: int, y: int) -> int:
    ret = 1
    tmp = x
    for i in range(30):
        if y & (1 << i):
            ret *= tmp
            ret %= MODULO
        tmp **= 2
        tmp %= MODULO
    return ret


a, b = map(int, input().split())
print(power(a, b))

File: rust/tessoku-book/src/bin/a29.rs
// A29 Power
// submission: https://atcoder.jp/contests/tessoku-book/submissions/38575451

use proconio::input;

const MODULO: u64 = 1_000_000_007;

fn power(x: u64, y: u64) -> u64 {
    let mut ret = 1;
    let mut tmp = x;
    for i in 0..30 {
        if y & (1 << i) > 0 {
            ret *= tmp;
            ret %= MODULO;
        }
        tmp = tmp * tmp;
        tmp %= MODULO;
    }
    return ret;
}

fn main() {
    input!{
        a: u64,
        b: u64,
    }
    println!("{}", power(a, b));
}
